Update documentation for initialize (#186)
precice/precice.github.io: pages/docs/couple-your-code/couple-your-code-initializing-coupling-data.md

summary: "As default values, preCICE assumes that all coupling variables are zero initially. For fluid-structure interaction, for example, this means that the structure is in its reference state. Sometimes, you want to change this behavior – for instance, you may want to restart your simulation."
---

{% version 3.0.0 %}

> `initializeData()` will be removed in preCICE version 3.0.0. You can do data initialization now by calling `initialize()`.

Reviewer: ```suggestion
`initializeData()` only exists for preCICE versions < 3.0. For versions >= 3.0, data initialization is directly handled in `initialize()`.
```

{% endversion %}

By default preCICE assumes that all coupling variables are zero initially. If you want to provide non-zero initial values, you can write data before calling `initialize()`. This data will then be used as initial data. The preCICE action interface that was introduced in [Step 6](couple-your-code-implicit-coupling) provides the following action to check whether initial data has to be written by a participant:

```cpp

> const std::string& constants::actionWriteInitialData()

Reviewer: Needs updates. "action interface" is gone.

Author: Done cdd2062 : **Important** we still need to update the main section about the actions. I don't want to update all the action things in this PR, because this is unrelated.

summary: "As default values, preCICE assumes that all coupling variables are zero initially. For fluid-structure interaction, for example, this means that the structure is in its reference state. Sometimes, you want to change this behavior – for instance, you may want to restart your simulation."
---

By default preCICE assumes that all coupling variables are zero initially. If you want to provide non-zero initial values, you can write data before calling `initialize()`. This data will then be used as initial data. To check whether initial data is required, you can use the following function:

```cpp

> bool requiresInitialData()

Author: This removal does not really belong to this documentation update, because it is triggered by the update of the action interface. But I did it here anyway.

void initializeData();
bool requiresInitialData()
```

Before jumping into the implementation, let's try to clarify how the usual the sequence of events in a serial and in a parallel coupling as studied in [Step 4](couple-your-code-coupling-flow) changes.

TODO: picture

In a serial coupling, only the second participant can send data inside `initializeData()`. In parallel coupling, both participants can initialize data.

The high-level API of preCICE makes it possible to enable this feature at runtime, irrelevant of serial or parallel coupling configuration. To support this feature, we extend our example as follows:
To support data initialization, we extend our example as follows:

```cpp
[...]

const std::string& cowid = precice::constants::actionWriteInitialData();

[...]

int displID = precice.getDataID("Displacements", meshID);
int forceID = precice.getDataID("Forces", meshID);
double* forces = new double[vertexSize*dim];

> double* displacements = new double[vertexSize*dim];

Author: This removal does not really belong to this documentation update, because it is triggered by the update of mesh and data ids. But I did it here anyway.

<exchange data="Displacements" mesh="StructureMesh" from="SolidSolver" to="FluidSolver" initialize="yes"/>
[...]
```

If you are using a serial coupling scheme, only initial data that is written from the second to the first participant is actually used. Initial data written from the first to the second participant will be ignored, because in a serial coupling scheme the first participant already computes its first result before the second participant even starts.

{% note %}
preCICE still supports data initialization for both participants, even if a serial coupling scheme is used. You can read our section on [time interpolation](couple-your-code-waveform) to learn more about the reasons.

> {% endnote %}

Reviewer: I had to read this part three times before understanding it. Could we improve it?
Instead of "written from A to B", I would always say "written to A and then communicated by preCICE from A to B".
Either all in a note or non. Remember that waveform will be the norm with v3 eventually.

Author: No so sure about this whole paragraph anymore, because it makes things more complicated and actually we are now planning to use initial data for every participant for every coupling scheme. Also the second participant in a serial coupling scheme uses the initial data it receives from the first. Even for waveform order = 0 we use the initial data for the relative time = 0. This is a very narrow edge case, but it leads to a consistent treatment across windows. See also the following figure:

![Screenshot from 2023-05-28 14-11-10](https://github.com/precice/precice.github.io/assets/5740604/2d424c5b-e724-476d-9074-f90148fa46f4)

How about removing this paragraph and opening the whole topic later when we introduce time interpolation?

Author: We could also use the figure from above later in the waveform relaxation section, if it is useful. For the section on data initialization, I think it contains too much information that has not been introduced yet.

Author: See [ec217b5](https://github.com/precice/precice.github.io/pull/186/commits/ec217b58ee30129a1d940147c7b2299b934af163) for some steps into this direction.
